### ISSUE TITLE
[Question] Build url

### ISSUE THREAD

**Issue by GioF71** [tamland/python-tidal#230]
Hello, I am upgrading the Tidal plugin for upmpdcli to the latest version 0.7.4 of your library.
Authentication mostly works, the navigation works perfectly just like with oauth2, but playback to the renderer fails.
My renderer is another instance of upmpdcli with mpd of course.

I don't know if this is related to [this issue](https://github.com/tamland/python-tidal/issues/226) of yours.

What I am currently doing is trying to decode the mpd (not MusicPD this time) manifest, and I get a list of url. But none of those actually play, neither in MPD or in any other player I tried, including VLC.

AFAIK MPD should be able to play MPEG-DASH (see [here](https://github.com/MusicPlayerDaemon/MPD/issues/560)) so I am probably doing something wrong.

Do you have any suggestion for this issue? 

Thank you

**Comment by tehkillerbee** [comment]
I have not yet looked into this but from my limited understanding, it is necessary to parse the MPEG-DASH manifest to create a playback URL suitable for your player. However, some players (ffmpeg) should be able to play directly.

Have you tried a different player?

https://stackoverflow.com/questions/47056177/how-to-play-mpeg-dash-audio-streams-from-console

**Comment by GioF71** [comment]
No I only tried MPD and VLC, but I will try it asap.
But how do you create that url which ends in .mpd?
Thanks a lot

**Comment by tehkillerbee** [comment]
I did just have success with this simple example, using the tidalapi.Stream as the input and then getting the URLs from the manifest.

```
manifest = json.loads(base64.b64decode(stream.manifest))
url =  manifest["urls"][0]
```
However, only one URL is returned and it's quality is not HI_RES

**Comment by GioF71** [comment]
Hello, thank you again, but it seems to me that s,manifest once b64decoded is xml, not json, am I doing something wrong?

**Comment by tehkillerbee** [comment]
Sorry, I noticed an error in the example. It should be 
```
f'data:application/dash+xml;base64,{base64.b64decode(stream.manifest)}'
```

**Comment by tehkillerbee** [comment]
The output should be an MPD string with an XML declaration header? That is what I got.

**Comment by GioF71** [comment]
Hello, this is the code I am trying. It is derived from the simple pkce login file.
I have put your code in a get_url method, but it goes in the "except" and does not return a URL.
Can you take a look?
Thank you!

```
import tidalapi
from tidalapi import Quality
from pathlib import Path
import os
import json
import base64

from mpegdash.parser import MPEGDASHParser

if os.path.exists("tidal-session-pkce.json"):
    print("file exists")
else:
    print("file does not exist")

def get_url(stream):
    try:
        manifest = json.loads(base64.b64decode(stream.manifest))
    except json.decoder.JSONDecodeError:
        return f'data:application/dash+xml;base64,{base64.b64decode(stream.manifest)}'
    return manifest["urls"][0]

session_file1 = Path("tidal-session-pkce.json")

session = tidalapi.Session()
# Load session from file; create a new session if necessary
session.login_session_file(session_file1, do_pkce=True)

# Override the required playback quality, if necessary
# Note: Set the quality according to your subscription.
# Low: Quality.low_96k
# Normal: Quality.low_320k
# HiFi: Quality.high_lossless
# HiFi+ Quality.hi_res_lossless
session.audio_quality = Quality.hi_res_lossless.value

album = session.album("110827651") # Let's Rock // The Black Keys
tracks = album.tracks()
# list album tracks
for track in tracks:
    print(track.name)
    # MPEG-DASH Stream is only supported when HiRes mode is used!
    s = track.get_stream()
    url = get_url(s)
    print(f"URL=[{url}]")

```

**Comment by GioF71** [comment]
Of course I have the json file with the credentials, so the login_session_file has success

**Comment by tehkillerbee** [comment]
Sorry, I edited my earlier example a while back.

You should change your get_url function to be the following
```python
def get_url(stream):
    try:
        manifest = json.loads(base64.b64decode(stream.manifest))
    except json.decoder.JSONDecodeError:
        return base64.b64decode(stream.manifest).decode('utf-8')
    return manifest["urls"][0]
```

**Comment by GioF71** [comment]
Hello, thank you very much, this code works.
The upmpdcli author tells me that I should serve the mpd url to MusicPD, but I don't know how to get that URL. Is that available through the library?

**Comment by tehkillerbee** [comment]
You should be able to feed music player daemon the manifest file for MPEG DASH streaming, i.e. the MPD XML,  if your player supports it. Have you tried saving it as an mpd file? 

This should also work if your ffmpeg installation has been built with the correct support:
```
```

If MPD expects a stream, you need to parse it further, eg to HLS format.
https://github.com/hyugogirubato/pydash2hls/tree/main

**Comment by GioF71** [comment]
Hello, I will try saving the mpd to a file and play it, anyway Music Player Daemon expects a stream, I cannot assume it to be on the same host as the upmpdcli instance or that it can anyway access the file

**Comment by GioF71** [comment]
Hello, I have successfully played the saved mpd file with ffmpeg, which is good!

However I couldn't generate the stream using the library. It fails when I load the xml manifest. Do you have a snippet of this conversion?

**Comment by tehkillerbee** [comment]
@GioF71 Sorry, I have not had time to look at it yet due to work. But I will try to get my conversion code added on a separate tidalapi branch so you can test with it.

**Comment by tehkillerbee** [comment]
Yes, what I have now is a M3U8 File for the HLS Stream. It can be loaded fine using ffmpeg (mp4 container with flac) and the bitrate is indeed 192khz. I would assume this HLS media playlist should work for you.

Ill clean up my code asap and get you a branch you can try to test with. I still need to figure out how to integrate all this with Mopidy :)

**Comment by GioF71** [comment]
Hello, thank you for sharing your progress. I look forward to enable the tidal plugin for upmpdcli to work with hires files!

**Comment by tehkillerbee** [comment]
@GioF71 I have added my latest changes in this branch. This is still a work in progress but I think that is enough to get you started:
https://github.com/tamland/python-tidal/tree/feature/hls-parsing

For mopidy-tidal, I decided to create a local HLS m3u8 playlist (file://xx) and pass it to mopidy instead of the file URL. Everything works as expected so far. Ofc. it requires the correct GStreamer plugins, i.e. ` sudo apt-get install gstreamer1.0-plugins-bad`.

**Comment by GioF71** [comment]
Hello, yes, I did not change anything other than the path of my credentials file!
Thank you!

**Comment by tehkillerbee** [comment]
@GioF71 Looks like the token refresh code was broken (I did get the same issue when the pkce session was somehow not valid anymore and I got a non MPEG-DASH stream). So this is the reason you get no playlist - there is no MPEG-DASH MPD available. 

I have fixed this on the before mentioned branch. Maybe you'll need to login again to make sure the correct access token is used.

**Comment by GioF71** [comment]
ok I have tried again, now the playlist are generated. 
However, both mpd and ffplay report errors:

MPD:

```
ffmpeg/mov,mp4,m4a,3gp,3g2,mj2: could not find corresponding trex (id 1)
ffmpeg/mov,mp4,m4a,3gp,3g2,mj2: could not find corresponding track id 0
ffmpeg/mov,mp4,m4a,3gp,3g2,mj2: trun track id unknown, no tfhd was found
ffmpeg/mov,mp4,m4a,3gp,3g2,mj2: error reading header
exception: Failed to decode https://sp-ad-fa.audio.tidal.com/mediatracks/some_chars/10.mp4?token=some_chars; avformat_open_input() failed: Invalid data found when processing input
```

ffplay:

```
[mov,mp4,m4a,3gp,3g2,mj2 @ 0xd6500630] could not find corresponding trex (id 1)
[mov,mp4,m4a,3gp,3g2,mj2 @ 0xd6500630] could not find corresponding track id 0
[mov,mp4,m4a,3gp,3g2,mj2 @ 0xd6500630] trun track id unknown, no tfhd was found
[mov,mp4,m4a,3gp,3g2,mj2 @ 0xd6500630] error reading header
https://sp-ad-fa.audio.tidal.com/mediatracks/some_chars/10.mp4?token=some_chars: Invalid data found when processing input
```

**Comment by GioF71** [comment]
If I try the playlist with ffplay, I get:

```
[hls @ 0xd6400630] Skip ('#EXT-X-VERSION:3')  0KB sq=    0B f=0/0
[hls @ 0xd6400630] Opening 'https://sp-ad-fa.audio.tidal.com/mediatracks/some_chars/0.mp4?token=some_chars' for reading
[https @ 0xd641d5a0] Protocol 'https' not on whitelist 'file,crypto,data'!
[hls @ 0xd6400630] Failed to open segment 0 of playlist 0
```

How do you play those playlists?

**Comment by GioF71** [comment]
tried to install but there is no difference :-(

**Comment by GioF71** [comment]
HA, I added `-protocol_whitelist file,http,https,tcp,tls,crypto` to ffplay command line, and now it plays. Is any of your selected tracks 24/192?

**Comment by GioF71** [comment]
set ffplay to play with alsa and it plays @192 kHz, that's really good progress. No clue how to convince mpd to play though
Thank you!

**Comment by tehkillerbee** [comment]
Yep, that must be added as well for that error to go away. Yep 192kHz/24bit playback works for me.
```
Input #0, hls, from 'dash_77646169_77646173.m3u8':sq=    0B f=0/0   
  Duration: 00:05:41.55, start: 0.000000, bitrate: 0 kb/s
...
  Stream #0:0: Audio: flac (fLaC / 0x43614C66), 192000 Hz, stereo, s32 (24 bit), 4645 kb/s (default)
    Metadata
    ....
```

**Comment by GioF71** [comment]
For playback to mpd, the author of upmpdcli told me I had to pass the url of the "MPD" (not music player daemon). Infact that played in ffmpeg. Is there a way to get that as a URL from your library?

**Comment by tehkillerbee** [comment]
Problem is you do not get an actual URL in the response from Tidal. In the response (json), you get a manifest that must be decoded to get you the MPD (MPEG-DASH). From that, the HLS media playlist can be constructed for playback.

But I assume MPD (the daemon) can work with playlists too? So perhaps you can just generate a local playlist and pass that instead? That is what I did with mopidy and that worked just fine.

**Comment by GioF71** [comment]
The problem is that I cannot write anything to mpd from upmpdcli (as a media server).
I might ask the author if I can serve a playlist. 
But even if this is possible, the problem would then be that for one song I would have a playlist (so multiple tracks). This breaks everything assumption upmpdcli (as a renderer) is based on.

**Comment by GioF71** [comment]
Anyway mpd does not decode the individual URLs anyway, so at the moment this is the first hurdle

**Comment by GioF71** [comment]
When I use BubbleUPnP own "Local and Cloud" with Tidal, the resulting URL for mpd is a plain flac. These is some processing on the BubbleUPnP app and/or BubbleUPnP server, that makes this possible. And of course the mpd+upmpdcli (renderer) stack in this case plays the track without any issue. It just 'sees' a regular flac over http.
I wonder if this is something I can implement in the tidal plugin on upmpdcli.

**Comment by tehkillerbee** [comment]
> When I use BubbleUPnP own "Local and Cloud" with Tidal, the resulting URL for mpd is a plain flac.

I assume BubbleUPnP only works for lower qualities - not HiRes? 

The tidalapi plugin can already give you a direct flac URL for playback, but only for <HiRes. For HiRes, only MPEG-DASH stream is available. If you try calling .get_url() when configured for HiRes playback, you will not get any URL. But when using low/high/lossless, you will get MP4 or FLAC (non HiRes of course).

**Comment by tehkillerbee** [comment]
> cache joined files 

I would expect this requires buffering before playback can be started so maybe that's not ideal. But I don't see why not. I think that is the same way it's done with some of the tidal downloader tools so perhaps you could take a look at them.

Or perhaps look at gapless upnp? https://forum.wiimhome.com/threads/gapless-mode-setnexturi-issues-for-upnp-interface.478/

**Comment by GioF71** [comment]
MPD already tries to acquire the next song to play in advance, iirc 20 seconds. So it can play gaplessly.
But for this to work properly, I would need to build the combined files in time, that might be challenging for long tracks.

**Comment by tehkillerbee** [comment]
What I mean is you already have a list of URLs for each track so perhaps they could be queued for (gapless) playback. Then you avoid buffering the whole track.

Out of curiosity, how does MPD/upmpdcli deal with live radio m3u8 (HLS) playlists with multiple segments? Surely they can be played back with MPD without any issues? 

It is the same principle here used by the playlist presented from tidalapi. It consists of segments, not individual tracks, so it should be handled as such by MPD (daemon), exactly as when playing back a radio stream. I am surprised if this is not possible to do with MPD, perhaps ask the developer for advice?

**Comment by GioF71** [comment]
Hello, I have tried 'serving' the m3u8 file via http, and VLC plays the whole track without a problem.
However, when I try to open the url using mpd, multiple items are enqueued instead of one, and anyway they won't play, because of a "unsupported uri scheme error".

MPD has a playlist plugin [here](https://mpd.readthedocs.io/en/stable/plugins.html#playlist-plugins) but m3u8 does not seem to be among the supported formats. 

Maybe that could be the issue, because if I try to add the first url in VLC, it doesn't work as well.
Is there a change to support any of the playlist formats that mpd supports?

**Comment by GioF71** [comment]
Hello, looks like the latest version on the branch you prepared fails with this error:

```
77646170: 'The Golden Age' by 'Beck'
MimeType:application/dash+xml
Traceback (most recent call last):
  File "/home/power/git/audio/python-tidal/examples/pkce_login.py", line 53, in <module>
    stream.bit_depth,
AttributeError: 'Stream' object has no attribute 'bit_depth'
```

I am using the version from a few commits before (c03864a117c3c309b1f1988d8da8c735aa2a9293, says "Added link to documentation" in the message, and this works.
Am I wrong? Thank you!

**Comment by tehkillerbee** [comment]
@GioF71 Hmm I cannot replicate this. Are you sure you are not using an older version of tidalapi but with the latest script?

I added the metadata in [cf42bc](https://github.com/tamland/python-tidal/commit/cf42bc2fcbcd545ee05a1df5d12ef38b0a51060f)

**Comment by GioF71** [comment]
Yes it's definitely possibile... I double checked but I must have done something wrong.
Anyway, is a release coming soon? :-)

**Comment by GioF71** [comment]
I confirm I was using a previous version in my python environment, this is solved now, thank you again

**Comment by GioF71** [comment]
Another thing, the upmpdcli author asks me if we can try and save (then serve) the mpd file instead of the m3u8 files, does it sound possible to you? Thank you again!

**Comment by tehkillerbee** [comment]
@GioF71 Sure, it's straightforward:
```python
stream = track.get_stream()
manifest = stream.get_stream_manifest()
data = manifest.get_manifest_data()
with open("{}_{}.mpd".format(album_id, track.id), "w") as my_file:
    my_file.write(data)
```

**Comment by GioF71** [comment]
Hello, thank you. I tried serving the mpd file instead of hls, and mpd (the player) works much better, allowing to seek the track, which did not work when using hls. I also see the bitdepth, sample rate and even bitrate (this is probably calculated by mpd because I did not provide it yet).

On the other hand, now Logitech Media Server refuses to play the stream. I can't find any log there. I see on the gui that the type is "application/octet-stream".
Do I need to set a different mimetype when serving the mpd?

**Comment by tehkillerbee** [comment]
Good to hear. Yes I switched to using the mpd manifest directly, gstreamer doesn't mind and it works well in mopidy-tidal.

Regarding logitech media server, I am not sure. A possible explanation could be that the mimetype is not dash+xml for that album, resulting in you passing on a non mpd manifest to LMS. I added an extra check to mopidy-tidal to avoid that problem.

What type of authentication are you using? Pkce?

**Comment by GioF71** [comment]
Yes, PKCE of course. With oauth2, LMS works. 
I also tried a WiiM pro as a UPnP renderer, and it's the same: it works with oauth2, not with pkce

**Comment by tehkillerbee** [comment]
@GioF71 Odd. Since it works with oauth2, it must be related to the MPD manifest. Have you tried stripping the XML header before passing on the manifest?

**Comment by GioF71** [comment]
Yes for sure it's related to the manifest... but I didn't try to strip xml header. But do you mean to remove just the following part?

`<?xml version='1.0' encoding='UTF-8'?>`

**Comment by tehkillerbee** [comment]
@GioF71 Sounds great. Good to hear you have made progress so far.

1. Regarding temporary mpd files - why do you save more than one file at a time? For caching purposes? 
2. Naturally :)
3. An official PyPi release is imminent, when testing has completed. It is very helpful that you have been testing it as well so we can iron out any bugs that I may have introduced.

How did you handle PKCE login?

**Comment by GioF71** [comment]
About question one: the plugin receives a request for the url of the next track a few seconds before it actually gets played. To that request I must answer with a mimetype and a url which can offer the mpd file, so I need to temporarily store the file. No caching, it's just a two-step process.

About pkce login, I have prepared python file which displays a generate json file as well as a set of env variables for the container.
If the user chooses the former, he just need to put the file in the correct location.
If the user chooses the latter, he just need to set those variables (same contents of the json file).

In the second case, the json file is created anyway and subsequently used.

The repo I have linked should make things even easier. The get-credential script will automatically place the file in the correct location, by running a docker run command using a mountpoint to the same directory that is used by the docker-compose file.

Now the challenge for me is to document the process so it can be easily understandable.

**Comment by GioF71** [comment]
About the file storing: I need to save the file because the http server is not controlled by the plugin, so I cannot create the file in real time.

**Comment by GioF71** [comment]
Hello again, can you confirm that getting the stream information (bit_depth, sample_rate etc) are requests that hit the playwall api? I get http 429 errors if I try to get these metadata too frequently. 
By too frequently, I mean for example, when I try to display any album with many tracks, let's say 40 is a critical number. 
Take this album for example: https://listen.tidal.com/album/112926479
Or am I doing something wrong?

**Comment by tehkillerbee** [comment]
See my answer here: https://github.com/tamland/python-tidal/issues/233 as this is not really related to this issue.

**Comment by tehkillerbee** [comment]
Closing this issue now, as I think we got all the issues sorted. Feel free to reopen if necessary.
